Issue/2708 Display downloadable files in the product details

hichamboushaba opened the pull request:
Fixes #2708

**Summary**
This PR allows displaying the downloadable files with their details on the product details fragment.

**Changes**

- Added a feature flag M4 to enable the feature
- Updated FluxC version to use the changes done on PR: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1663
- Updated the card of downloadable files in the list of product details, when M4 feature is enabled, the old card won't be shown too.
- Added FragmentProductDownloads to display the list of files.
- Fixed the test `Displays the product detail properties correctly`, and updated it to confirm the handling of downloadable files correctly.

**Screenshots**
![Screen Shot 2020-08-13 at 19 39 11](https://user-images.githubusercontent.com/1657201/90173663-b6d0df00-dd9c-11ea-9ec3-efd1810853d1.png) ![Screen Shot 2020-08-13 at 19 39 20](https://user-images.githubusercontent.com/1657201/90173669-b89aa280-dd9c-11ea-9580-8bbdb7ff3002.png)

**Testing**
- Make sure that one of the products in the store is downloadable, and has some files added in wp-admin.
- Go to the product details.
- Notice the downloadable files section.
- Click on the downloadable files section to check the details of each file.

hichamboushaba committed:
add a feature flag for release m4

hichamboushaba committed:
update fluxC's version

hichamboushaba committed:
update Product model to expose the downloadable files

hichamboushaba committed:
display the downloadable files in the product cards

hichamboushaba committed:
Add a fragment to display the list of downloadable files

hichamboushaba committed:
add exit event

hichamboushaba committed:
fix product details cards test

hichamboushaba committed:
Add a test case for downloadable files

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailCardBuilder.kt:
```diff
@@ -95,7 +96,8 @@ class ProductDetailCardBuilder(
                 product.tags(),
-                product.shortDescription()
+                product.shortDescription(),
+                product.downloads()
```
In Figma designs, the downloadable files section is not last, but the whole ordering is different than what we have now.
So I decided to put it in the last position for now, until confirming with the design team.

anitaa1990 replied:
I believe this is the ordering we have in Figma and in the app for Simple Products 🤔 I seee reviews are missing from the screenshot but only because they are only displayed if reviews are enabled and it is displayed below the product type. 

Are you seeing a different order for Simple products?

<img width="105"  src="https://user-images.githubusercontent.com/22608780/90219727-54232600-de24-11ea-8b98-854d7a9260ff.png">

The downloadable files option comes above Short description and below Product tags.

hichamboushaba replied:
Yes, I see a different ordering in Figma:
<img width="202" alt="Screen Shot 2020-08-14 at 09 29 23" src="https://user-images.githubusercontent.com/1657201/90229980-ba9e4900-de10-11ea-9404-b10d84d83c85.png">

Until we confirm the cause of this discrepancy, I'll move the downloads section above short description.

anitaa1990 replied:
I see a different order in separate screens as well 😕 I agree that we can wait to fix the order till @Garance91540 comes back next week 👍

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductNavigationTarget.kt:
```diff
@@ -49,4 +49,5 @@ sealed class ProductNavigationTarget : Event() {
     data class ViewProductDetailBottomSheet(val remoteId: Long) : ProductNavigationTarget()
     data class ViewProductTypes(val remoteId: Long) : ProductNavigationTarget()
     data class ViewProductReviews(val remoteId: Long) : ProductNavigationTarget()
```
I noticed we are passing the `remoteId` as argument to many fragments, but the argument is not being read by the destination fragment.
Is this used for something?

anitaa1990 replied:
Good find! I think there is some work being done by Ondrej for product variations that might still use the passed `remoteId`.

hichamboushaba committed:
Fix ktlint issues

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/util/StringUtils.kt:
```diff
@@ -46,6 +47,34 @@ object StringUtils {
+     * @param [zero] Optional. The desired string identifier to use when [quantity] is exactly 0.
+     * @param [one] Optional. The desired string identifier to use when the [quantity] is exactly 1
+     */
+    fun getQuantityString(
+        resourceProvider: ResourceProvider,
+        quantity: Int,
+        @StringRes default: Int,
+        @StringRes zero: Int? = null,
+        @StringRes one: Int? = null
+    ): String {
+        return when (quantity) {
+            0 -> resourceProvider.getString(zero ?: default, quantity)
+            1 -> resourceProvider.getString(one ?: default, quantity)
+            else -> resourceProvider.getString(default, quantity)
+        }
+    }
+
```
Nice <3 Love the detailed comments here 😍

anitaa1990 commented on WooCommerce/src/main/res/layout/product_downloads_list_item.xml:
```diff
@@ -0,0 +1,54 @@
+
+    <com.google.android.material.textview.MaterialTextView
+        android:id="@+id/name"
+        style="@style/Woo.ListItem.Title"
```
Can we change the names of the ids (`icon`, `name`, `url`) here to specify that it's used for the product downloadable files screen? something like `product_download_item_icon`?

anitaa1990 commented on WooCommerce/src/main/res/layout/fragment_product_downloads_list.xml:
```diff
@@ -0,0 +1,50 @@
+                app:layout_constraintTop_toTopOf="parent" />
+        </androidx.constraintlayout.widget.ConstraintLayout>
+
+    </com.google.android.material.card.MaterialCardView>
```
I believe there is a custom class defined to add a new button to a screen. We can reuse the same component here rather than adding a new layout? See [add_product_category_view](https://github.com/woocommerce/woocommerce-android/blob/9eb5a767b5079ff41009cd81967d0c7bc0e67a1d/WooCommerce/src/main/res/layout/add_product_category_view.xml) and [AddProductCategoryView](https://github.com/woocommerce/woocommerce-android/blob/0c65d2fc10988a8133e11cd5e60444609ae4fa1c/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/categories/AddProductCategoryView.kt) class (Maybe change the name of the classes so that it's generic and reusable 😉 ).

anitaa1990 reviewed (commented):
Nice work @hichamboushaba! The changes look good and everything works as expected! I left a few code comments for your reference and updated the base branch to a feature branch 👍 

One small feedback I'd like to give is to add some testing steps in the PR description so that others not associated with the project might be quickly able to test the given scenarios without needing any additional input from our end 😄 

Also, this is just a nit from my end that, there are a lot of unrelated alignment changes in the PR that make it a bit harder to review the actual changes associated with the PR.

hichamboushaba committed:
Rename the ids of download list item

hichamboushaba committed:
Reuse the add view from the categories screen

hichamboushaba committed:
Move the downloadable files section above short description

hichamboushaba committed:
Code cleanup

hichamboushaba commented:
@anitaa1990 thanks for the review, I addressed your comments, please let me know if I missed anything 😉 .

> Also, this is just a nit from my end that, there are a lot of unrelated alignment changes in the PR that make it a bit harder to review the actual changes associated with the PR.

Yes, you are right, I'm used to using the format document shortcut when editing, and I don't know if my editor settings are different than the project's defaults or what, but this caused these changes.
I'll see what's causing it, and I'll make sure it doesn't happen on next PRs 🙂 

> One small feedback I'd like to give is to add some testing steps in the PR description so that others not associated with the project might be quickly able to test the given scenarios without needing any additional input from our end 😄

I added testing steps to the PR's description 🙂

anitaa1990 reviewed (approved):
LGTM :shipit:
